### PR TITLE
Av parser

### PR DESCRIPTION
**Pull request by StankAnt**

### CODE REVIEW

File: spec/avParser.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  it('should be return an array of marks', async () => {
+    requestStub = sinon.stub(request, 'get').callsFake((opts: any) => parser.MARKS);
+    await api.updateMarks();
+    assert.deepEqual(api.getMarks() as any, parser.MARKS_RESULT);
+  });
+
+  it('should be return an array of models', async () => {
+    requestStub = sinon
+      .stub(request, 'get')
+      .callsFake((opts: any) => new Promise(resolve => resolve(parser.MODELS)));
```
Comment (AntonNarkevich) [inline spec/avParser.ts]: Please use async function here. :) You gonna love it.

File: src/parsers/avParser/index.ts [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    });
+  };
+
+  let models: any[] = [];
+  const date = new Date();
+  for (let i = 0; i < marks.length; i += 3) {
+    const currentMarks = _.slice(marks, i, i + 3);
+    models = _.concat(
+      models,
+      await Promise.all(_.slice(
```
Comment (AntonNarkevich) [inline src/parsers/avParser/index.ts]: Please either make it more readable or add some comments. Right now it's not clear what happens here.

File: src/parsers/avParser/index.ts [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+      },
+      url: AV_URL
+    });
+    const $ = cheerio.load(response, {
+      normalizeWhitespace: true,
+      xmlMode: true
+    });
+    const bodyTypes: string[] = $('#body_id option')
+      .map(function() {
+        return $(this).text();
```
Comment (AntonNarkevich) [inline src/parsers/avParser/index.ts]: Arrow function here please :0

File: src/parsers/avParser/index.ts [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        xmlMode: true
+      });
+      const adsURLs: string[] = $('.listing-item-image-in')
+        .find('a')
+        .map(function() {
+          return $(this).attr('href');
+        })
+        .get();
+
+      const get = (url: any) => {
```
Comment (AntonNarkevich) [inline src/parsers/avParser/index.ts]: Please don't combine then's and awaits. Use awaits only.

File: src/utils/parserUtils.ts [hunk 5]
```diff
@@ -97,3 +97,63 @@ export const updateServiceData = async () => {
...
   await updateDBData(transfomedMarks, models, bodyTypes);
 };
+
+export const transformAvByBodyTypes = (bodyTypes: any[]) => {
+  return _.chain(bodyTypes)
+    .map(type => {
+      const name = _.capitalize(type);
+      return name === 'Легковой фургон'
+        ? name
+        : _.chain(name)
```
Comment (AntonNarkevich) [inline src/utils/parserUtils.ts]: Please add some comments here.

**Review by AntonNarkevich — COMMENTED** [review]
I left some comments, please take a look.
Let's make a second pass then.

File: spec/liquidityCalculator.ts [hunk 6]
```diff
@@ -49,6 +49,7 @@ describe('Calculate liquidity', () => {
         creationDate: moment('2016-11-10'),
         isSold: true,
         kms: adFields.KMS,
+        lastTimeUpDate: moment('2016-11-10'),
```
Comment (AntonNarkevich) [inline spec/liquidityCalculator.ts:52]: moment call returns a moment object not a date. is it OK for us in this case? shouldn't we call .toDate() here?

File: spec/onlinerParser.ts [hunk 7]
```diff
@@ -98,49 +92,27 @@ describe('Onliner Parser', () => {
     MODELS_RESULT: { '183': [{ "'33 Hot Rod": '2581' }, { '818': '2583' }] }
   };
 
-  let puppeteerStub: sinon.SinonStub;
   let requestStub: sinon.SinonStub;
 
   afterEach(done => {
```
Comment (AntonNarkevich) [inline spec/onlinerParser.ts]: Callback? Seriously? Please do async function here.

File: spec/onlinerParser.ts [hunk 8]
```diff
@@ -98,49 +92,27 @@ describe('Onliner Parser', () => {
...
 
   it('should be return an array of body types', async () => {
-    puppeteerStub = sinon.stub(puppeteer, 'launch').returns({
-      close: () => '',
-      newPage: () => ({
-        content: () => parser.BODY_TYPES,
-        goto: (url: string, options: any) => ''
-      })
-    });
+    requestStub = sinon.stub(request, 'get').callsFake((opts: any) => parser.BODY_TYPES);
```
Comment (AntonNarkevich) [inline spec/onlinerParser.ts:114]: Where is a a restore for this 'stub'?

File: src/index.ts [hunk 9]
```diff
@@ -10,25 +10,36 @@ import * as logger from 'koa-logger';
...
+import { updateAvByData, updateOnlinerData } from './utils/parserUtils';
+import { torTriggerer } from './utils/torTriggerer';
 
-// tslint:disable-next-line:no-var-requires
-const cors = require('@koa/cors');
+import { sendMessageToSlack } from './utils/slack';
 
-(mongoose as any).Promise = bluebird;
+import { Api } from './parsers';
+global.Promise = bluebird;
```
Comment (AntonNarkevich) [inline src/index.ts]: do we need this? Shouldn't Bluebird use a native promise internally?

File: src/parsers/avParser/index.ts [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+import { AV_URL, codeErrors, proxy } from '../../config/config';
+import { IAvMark } from '../../interfaces/parserInterface';
+import { ParserError } from '../../utils/errors/';
+import { transformBmvAvModel, transformMercedesAvModel } from '../../utils/parserUtils';
+import { torTriggerer } from '../../utils/torTriggerer';
+
+// tslint:disable-next-line:no-var-requires
+const Agent = require('socks5-https-client/lib/Agent');
+// tslint:disable-next-line:no-var-requires
+const _ = require('lodash');
```
Comment (AntonNarkevich) [inline src/parsers/avParser/index.ts:13]: why are we doing require?

File: src/parsers/avParser/index.ts [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+  let response: any;
+  try {
+    response = await sendRequest(model.url);
+    let $ = cheerio.load(response, {
+      normalizeWhitespace: true,
+      xmlMode: true
+    });
+
+    const pagesCount =
+      +$('.pages-arrows-index')
```
Comment (AntonNarkevich) [inline src/parsers/avParser/index.ts]: Please use parseInt instead of this tricky conversion.

File: src/services/index.ts [hunk 12]
```diff
@@ -24,6 +34,9 @@ import {
   updateUserProfile
 } from './userService';
 
+// tslint:disable-next-line:no-var-requires
+const _ = require('lodash');
```
Comment (AntonNarkevich) [inline src/services/index.ts:38]: stop requiring please, use imports

File: src/utils/parserUtils.ts [hunk 13]
```diff
@@ -1,24 +1,25 @@
...
+import { updateDBData, updateDBDataFromOnliner, updateDBDateFromAvBy } from '../services/';
 import { getBodyTypeByName } from '../services/bodyTypeService';
 import { getMarkByName } from '../services/markService';
 import { updateMarks } from '../services/markService';
-import { getModelByName, saveNewModel } from '../services/modelService';
+import { getModelByName, getModelByNameAndMarkId, saveNewModel } from '../services/modelService';
 
-export const transformOnlinerDate = (onlinerDate: string): string => {
+// tslint:disable-next-line:no-var-requires
+const _ = require('lodash');
```
Comment (AntonNarkevich) [inline src/utils/parserUtils.ts:12]: import

File: src/utils/parserUtils.ts [hunk 14]
```diff
@@ -110,5 +111,82 @@ export const updateServiceData = async () => {
...
   const bodyTypes = api.getBodyTypes();
-  await updateDBData(transfomedMarks, models, bodyTypes);
+  await updateDBDataFromOnliner(transfomedMarks, models, bodyTypes);
+};
+
+export const transformAvByBodyTypes = (bodyTypes: any[]) => {
+  return _.chain(bodyTypes)
+    .map((type: any) => {
+      const name = _.capitalize(type);
+      return name === 'Легковой фургон' // we take only first word of the bodytype name, exclude "Легковой фургон"
```
Comment (AntonNarkevich) [inline src/utils/parserUtils.ts:121]: exclude Легковой фургон. Sounds really nice!

**Review by AntonNarkevich — CHANGES_REQUESTED** [review]

File: spec/onlinerParser.ts (hunk 8 above)
Comment (StankAnt) [reply spec/onlinerParser.ts:114]: In afterEach hook

File: spec/liquidityCalculator.ts (hunk 6 above)
Comment (StankAnt) [reply spec/liquidityCalculator.ts:52]: It works correctly

File: src/parsers/avParser/index.ts (hunk 10 above)
Comment (StankAnt) [reply src/parsers/avParser/index.ts:13]: Types for lodash don't work correctly, "Agent" types doesn't exists. And a linter swears

File: src/services/index.ts (hunk 12 above)
Comment (StankAnt) [reply src/services/index.ts:38]: Doesn't work without it

File: src/utils/parserUtils.ts (hunk 13 above)
Comment (StankAnt) [reply src/utils/parserUtils.ts:12]: The same

**Pull request merged**